server.py

```diff
@@ -1,36 +1,376 @@
+            "gender",
+            "languages",
+            "need",
+            "rights",
+            "housing",
+        ],
+    )
+):
+    """A coach object - an entry in the database."""
+
+    def __new__(
+        cls,
+        *,
+        id: Optional[int] = None,
+        name: str,
+        bio: str,
+        available: bool = True,
+        birth_year: int,
+        gender: str,
+        languages: Dict[str, int],
+        need: Collection[int],
+        rights: Collection[int],
+        housing: Collection[int],
+    ):
+        # TODO: Validate args.
+        return super().__new__(
+            cls,
+            id=id,
+            name=name,
+            bio=bio,
+            available=available,
+            birth_year=birth_year,
+            gender=gender,
+            languages=languages,
+            need=need,
+            rights=rights,
+            housing=housing,
+        )
+
+    def to_json(self) -> CoachJSON:
+        return self._asdict()
 
-def check_request_for_authorization_status(request):
-    sanic.log.logger.info(f"Checking authorization: IP is {request.ip}")
-    match = (request.ip == "127.0.0.1")
-    sanic.log.logger.info(f"Check IP matches 127.0.0.1: {match}")
+    @classmethod
+    def from_json(cls, obj: CoachJSON) -> "Coach":
+        return cls(**obj)
 
-    return match
+    def copy(self) -> "Coach":
+        return self.from_json(self.to_json())
 
-def authorized(f):
-    @functools.wraps(f)
-    async def decorated_function(request, *args, **kwargs):
-        is_authorized = check_request_for_authorization_status(request)
 
-        if is_authorized:
-            # Run the handler method and return the response.
-            response = await f(request, *args, **kwargs)
-            return response
+class CoachDB:
+    """The in-memory coach database."""
+
+    def __init__(
+        self, file: Optional[os.PathLike] = None, *, auto_persist: bool = True
+    ):
+        """
+        :param file:
+            File to use as persistent storage of the DB. If not found it will be
+            created.
+        :param auto_persist:
+            Whether to automatically persist the DB to file on operations that
+            modify the DB. Ignored if file is not given.
+            Set this to False if manual batching is desired.
+        """
+        self._db: Dict[int, Coach] = dict()
+        self._file = file
+        self._auto_persist = auto_persist and file
+        self._next_id: int = 0
+        if file:
+            self._load_from_file(file)
+
+    def __iter__(self):
+        return iter(self._db.values())
+
+    def __contains__(self, item: Union[int, Coach]):
+        if item in self._db:
+            return True
+        elif isinstance(item, Coach):
+            return item.id in self._db
+        else:
+            return False
+
+    def add(self, coach: Coach) -> Coach:
+        """Add a new coach to the DB, setting the ID."""
+        assert self._next_id not in self._db
```

Comment:
This looks like it would be hit if you restart the server and then add a coach - see earlier comment